---
title: "Lab 1: Gene expression microarrays and differential expression"
author: "Leslie Myint, Jean-Philippe Fortin"
date: "March 29, 2016"
package: "`r pkg_ver('BiocStyle')`"
output: 
  BiocStyle::html_document
---

# Overview

We will be looking at data from a study of gene expression in chronic lymphocytic leukemia (CLL). The authors were interested in identifying genes that are differentially expressed in stable versus progressive cancers.

# Reading your own data

Helpful information can be found in the [affy vignette](http://bioconductor.org/packages/release/bioc/vignettes/affy/inst/doc/affy.pdf) and the [affy reference manual](http://bioconductor.org/packages/release/bioc/manuals/affy/man/affy.pdf). Let's say that I have .CEL files downloaded to a folder called "microarray" on my Desktop.

```{r eval=FALSE}
# Navigate to the folder containing the .CEL files
setwd("~/Desktop/microarray/")
# Read in the data using...
abatch <- ReadAffy() # this reads in all .CEL files in this directory and stores in the abatch object
# ...or...
files <- list.files(".") # get filenames of every file in this directory
abatch <- read.affybatch(filenames = files)
```

# Loading packages and data

If you don't already have the packages installed:

```{r eval=FALSE}
source("https://bioconductor.org/biocLite.R")
biocLite(c("CLL", "affy", "limma", "hgu95av2.db"))
install.packages("scales")
```

Load the packages:

```{r warning=FALSE, message=FALSE}
library(CLL)
library(affy)
```

What forms of data are available to us?

```{r}
data(package = "CLL")
```

We see that there are both raw and processed forms of data. **CLLbatch** is a raw data object. **disease** contains metadata information. **sCLLex** is an **ExpressionSet** object which means that it has gone through some processing. There are also some analysis results that we won't be using. Let's get the first objects into our R workspace:

```{r}
data(CLLbatch, sCLLex, disease, package = "CLL")
```

# Exploratory data analysis

We can get some general information about the experiment by printing the **AffyBatch** object:

```{r warning=FALSE, message=FALSE}
CLLbatch # same as print(CLLbatch)
```

What sample classes do the arrays belong to?

```{r}
disease
```

For ease of exploration and analysis later, let's store the samples that belong to the two classifications of CLL (stable and progressive):

```{r}
samplesProg <- which(disease$Disease=="progres.")
samplesStable <- which(disease$Disease=="stable")
```

## Microarray images

How do the microarray images look? Are there any obvious spatial artifacts?

```{r fig.width=8, fig.height=12}
# Plot all 24 of the array images (can be slow)
par(mfrow = c(6,4), mar = c(2,2,2,1))
image(CLLbatch)
```

```{r fig.width=8, fig.height=4}
# Plot the first 4 progressive and the first 4 stable samples
par(mfrow = c(2,4), mar = c(2,2,2,1))
image(CLLbatch[, c(samplesProg[1:4], samplesStable[1:4])])
```

## Intensity distributions

We can also look at the distributions of log-intensities across the arrays:

```{r}
par(mfrow = c(1,1), mar = c(4,4,3,1))
hist(CLLbatch)
boxplot(CLLbatch)
```

From the density estimates, we see a typical shape that resembles the mixture of a normal and an exponential distribution.

## MA plots

MA plots of individual arrays versus individual arrays:

```{r fig.width=8, fig.height=8, warning=FALSE, message=FALSE}
par(mfrow = c(4,4), mar = c(4,4,3,1))
MAplot(CLLbatch[, c(samplesProg[1:2], samplesStable[1:2])], pairs = TRUE, plot.method = "smoothScatter")
```

We can make an MA plot of one group versus the other using the perfect match (PM) data in the **CLLbatch** object. First let's get acquainted with the object containing the PM data.

```{r}
pmdata <- pm(CLLbatch)
head(pmdata)
dim(pmdata)
all.equal(paste0(disease$SampleID, ".CEL"), colnames(pmdata))
```

We see that there are ``r nrow(pmdata)`` probes (rows) and that the ordering of the samples in the **disease** metadata object is the same as in the columns of the **pmdata** object. Now we can continue with the calculations needed to make the group vs. group MA plot. Instead of comparing one sample's expression levels to another sample's expression levels, we will compare the average expression levels in one group to those in the other group.

```{r fig.width=8}
logpmdata <- log2(pmdata + 1)
pmdataProg <- logpmdata[,samplesProg]
pmdataStable <- logpmdata[,samplesStable]
meanProg <- rowMeans(pmdataProg)
meanStable <- rowMeans(pmdataStable)
A <- (meanProg + meanStable)/2
M <- meanProg - meanStable
library(scales) # Need the scales package for alpha blending
par(mfrow = c(1,2))
plot(A, M, pch = 16, col = alpha("black", 0.1), xlab = "Average log2 expression", ylab = "Log2 expression ratio", main = "Progressive vs. stable")
abline(h = 0, lwd = 3, col = "deepskyblue")
lo <- lowess(A, M, f = 0.2)
lines(lo, lwd = 3, col = "deeppink")
# Can also use smoothScatter
smoothScatter(A, M, xlab = "Average log2 expression", ylab = "Log2 expression ratio", main = "Progressive vs. stable")
lines(lo, col = "deeppink", lwd = 3)
abline(h = 0, col = "deepskyblue", lwd = 3)
```

The cloud is clearly not centered about the $y = 0$ line, but this might be what we expect in cancer.

## Probe(set) names

Each row of **pmdata** corresponds to a probe, and each probe belongs to a probeset (a set of probes targeting the same gene). We can get the probeset names with **geneNames** and the probeset that each probe belongs to with **probeNames**.

```{r}
pn <- probeNames(CLLbatch)
gn <- geneNames(CLLbatch)
identical(unique(pn), gn)
length(pn)==nrow(pmdata)
head(pn) # The first few probes all belong to the same probeset
head(gn)
```

More information about the syntax for these probeset names can be found here: [http://www.affymetrix.com/support/help/faqs/mouse_430/faq_8.jsp](http://www.affymetrix.com/support/help/faqs/mouse_430/faq_8.jsp). We can obtain a subset of the perfect match (PM) or mismatch (MM) intensities corresponding to certain genes (probesets):

```{r}
pm(CLLbatch, gn[1])
mm(CLLbatch, gn[c(1,3)])
```

# Running processing algorithms

If you want to run RMA or MAS5 (the Affymetrix method), the quickest way to get a processed data object is to run:

```{r eval=FALSE}
eset <- rma(CLLbatch)
eset <- mas5(CLLbatch)
```

To control exactly what methods are used at each stage of processing, we can use **expresso**:

```{r warning=FALSE}
eset <- expresso(CLLbatch, normalize.method = "quantiles", bgcorrect.method = "rma", pmcorrect.method = "pmonly", summary.method = "liwong")
```

More information on different methods for normalization (**normalize.method**), background correction (**bgcorrect.method**), probe-specific background correction (**pmcorrect.method**), and probe set summarization (**summary.method**) can be found in section 3.3 of the [affy overview vignette](https://bioconductor.org/packages/release/bioc/vignettes/affy/inst/doc/affy.pdf) and in the [affy built-in methods vignette](https://bioconductor.org/packages/release/bioc/vignettes/affy/inst/doc/builtinMethods.pdf).

In particular calling the **rma** function is equivalent to using **normalize.method = "quantiles", bgcorrect.method = "rma", pmcorrect.method = "pmonly", summary.method = "medianpolish"** and calling the **mas5** function is equivalent to using **normalize.method = "constant", bgcorrect.method = "mas", pmcorrect.method = "mas", summary.method = "mas"**.

## Evaluating impact of processing

The **eset** object is an **ExpressionSet** object containing metadata information and processed expression values. We can extract the processed expression values with **exprs**:

```{r}
processed <- exprs(eset)
head(processed)
dim(processed)
nrow(processed)==length(gn)
```

Both boxplots and density plots show that processing causes the intensity distributions to be much more similar.

```{r fig.width=8}
par(mfrow = c(1,2))
boxplot(CLLbatch, main = "Before processing")
boxplot(log2(processed + 1), main = "After processing")
```

```{r fig.width=8}
par(mfrow = c(1,2))
hist(CLLbatch, main = "Before processing")
plotDensity(log2(processed + 1), main = "After processing")
```

We also see that the MA plot is more symmetric about $y = 0$.

```{r fig.width=8}
A.pre <- rowMeans(logpmdata)
M.pre <- rowMeans(logpmdata[,samplesProg]) - rowMeans(logpmdata[,samplesStable])
par(mfrow = c(1,2))
plot(A.pre, M.pre, pch = 16, col = alpha("black", 0.1), xlab = "Average log2 expression", ylab = "Log2 expression ratio", main = "Progressive vs. stable (before)", ylim = c(-2,2))
abline(h = 0, lwd = 3, col = "deepskyblue")
lines(lowess(A.pre, M.pre, f = 0.2), lwd = 3, col = "deeppink")
A.post <- rowMeans(log2(processed + 1))
M.post <- rowMeans(log2(processed[,samplesProg] + 1)) - rowMeans(log2(processed[,samplesStable] + 1))
plot(A.post, M.post, pch = 16, col = alpha("black", 0.1), xlab = "Average log2 expression", ylab = "Log2 expression ratio", main = "Progressive vs. stable (after)", ylim = c(-2,2))
abline(h = 0, lwd = 3, col = "deepskyblue")
lines(lowess(A.post, M.post, f = 0.2), lwd = 3, col = "deeppink")
```

# Differential expression analysis

**Goal:** Find genes that have statistically significant different abundances between two (or more) groups. In this case, the two conditions are progressive and stable chronic lymphocytic leukemia. We will use the package **limma** for analysis, and we will use the processed data from the original study which is contained in the **sCLLex** object.

```{r warning=FALSE, message=FALSE}
library(limma)
eset <- sCLLex
```

## Linear model setup

For each gene, we want to fit the model $y_{i} = \beta_0 + \beta_1 x_i$ where $y_i$ is the gene's expression value for sample $i$, $\beta_0$ is an intercept, $x_i$ is a dummy variable for group, and $\beta_1$ is the coefficient associated with the change in expression between the two groups. Our goal is to test $H_0: \beta_1 = 0$ for each gene.

## Create design matrix

We extract the phenotype information and use this to construct our linear regression design matrix.

```{r}
pheno <- pData(eset)
groups <- as.factor(pheno$Disease)
designMat <- model.matrix(~groups)
colnames(designMat) <- c("Intercept","Disease")
designMat
```

## Linear model fitting

We can fit the linear models using the **lmFit** function in **limma**. All we need is an expression matrix containing the summarized intensities and the design matrix:

```{r}
fit <- lmFit(object = eset, design = designMat)
```

**fit** is an object containing the results of the fits. For instance, we can retrieve the coefficients of the linear model with

```{r}
head(fit$coefficients)
```

## Empirical Bayes variance estimation

Computing classical t-statistics for experiments with a small number of arrays is unstable. The empirical Bayes method implemented in **limma** borrows information about expression variability across genes to make the analysis more stable. Specifically, the function **eBayes** computes moderated t-statistics by estimating the standard errors using empirical Bayes moderation. 

```{r}
fit2 <- eBayes(fit)
names(fit2)
```

## Results

Let's look at the distribution of p-values:

```{r}
par(mfrow = c(1,1))
hist(fit2$p.value[,2], xlab = "p-value", main = "")
```

A **volcano plot** is used to plot significance (e.g.instance p-value) versus fold-change (e.g. diff. expression coefficient):

```{r}
p <- fit2$p.value[,2]
plot(fit2$coefficients[,2], -log2(p), pch = 20, cex = 0.3, xlab = "Log fold change")
abline(v = c(-1,1), lwd = 3, col = "darkorange")
abline(h = 5, lwd = 3, col = "darkorange")
```

**limma** has a command to produce a volcano plot on a fitted object produced by the **eBayes** command:

```{r}
volcanoplot(fit2, coef = 2)
```

## Multiple testing adjustment

Because we are testing thousands of hypotheses, we need to correct our p-values for this multiplicity of tests. **p.adjust()** returns adjusted p-values for different methods:

```{r}
b.pvalues <- p.adjust(p, method = "bonferroni")
summary(b.pvalues)
head(sort(b.pvalues))
fdr.pvalues <- p.adjust(p, method = "fdr")
summary(fdr.pvalues)
head(sort(fdr.pvalues))
```

We can also use the function **topTable** in the **limma** package to return the top differentially expressed genes. For instance, to return the top 100 genes that pass an uncorrected p-value threshold of $0.05$, we can use

```{r}
table <- topTable(fit2, coef = 2, adjust.method = "none", p.value = 0.05, number = 100)
head(table)
```

# Biological significance

We still don't have the names of the genes associated with our probesets. We only have probset IDs. To map these probesets to genes, we need first to load the annotation package containing the relevant information for this Affymetrix array: 

```{r, warning=FALSE, message=FALSE}
library(hgu95av2.db)
columns(hgu95av2.db) # To see the info stored in the annotation
```

To find the gene symbol corresponding to the top 100 genes:

```{r, warning=FALSE, message=FALSE}
probeNames <- rownames(table) # Names of the top 100 probes
info <- select(hgu95av2.db, probeNames, c("SYMBOL"), "PROBEID")
# Other names from columns(hgu95av2.db) can be added to the c("SYMBOL") above:
# info <- select(hgu95av2.db, probeNames, c("SYMBOL", "ENTREZID"), "PROBEID")
head(info)
dim(info)
```

Wait, we had ``r nrow(table)`` probes, and now we have ``r nrow(info)`` gene symbols. What's wrong? It seems that some probesets map to more than one gene:

```{r}
table(table(info$PROBEID))
```

93 probesets map to 1 gene, 6 probesets map to 2 genes, and 1 probeset maps to 5 genes. Let's use only the first gene mapped for each probeset:

```{r}
info <- info[match(probeNames, info$PROBEID),]
dim(info)
```

We could have also decided to remove these probesets or something else depending on our downstream analysis plan. Add the gene symbol information to our table:

```{r}
table <- cbind(table, info$SYMBOL)
head(table)
```